nifi-nar-bundles/nifi-extension-utils/nifi-processor-utils/src/main/java/org/apache/nifi/processor/util/list/ListedEntityTracker.java

```diff
@@ -0,0 +1,322 @@
+        }
+
+        final long currentTimeMillis = CURRENT_TIMESTAMP.get();
+        final long watchWindowMillis = context.getProperty(TRACKING_TIME_WINDOW).evaluateAttributeExpressions().asTimePeriod(TimeUnit.MILLISECONDS);
+
+        final String initialListingTarget = context.getProperty(INITIAL_LISTING_TARGET).getValue();
+        final long minTimestampToList = (initialListing && INITIAL_LISTING_TARGET_ALL.getValue().equals(initialListingTarget))
+                ? -1 : currentTimeMillis - watchWindowMillis;
+
+        final Collection<T> listedEntities = listEntities.apply(minTimestampToList);
+
+        if (listedEntities.size() == 0) {
+            logger.debug("No entity is listed. Yielding.");
+            context.yield();
+            return;
+        }
+
+        final List<T> updatedEntities = listedEntities.stream().filter(entity -> {
+            final String identifier = entity.getIdentifier();
+
+            if (entity.getTimestamp() < minTimestampToList) {
+                logger.trace("Skipped {} having older timestamp than the minTimestampToList {}.", new Object[]{identifier, entity.getTimestamp(), minTimestampToList});
+                return false;
+            }
+
+            final ListedEntity alreadyListedEntity = alreadyListedEntities.get(identifier);
+            if (alreadyListedEntity == null) {
+                logger.trace("Picked {} being newly found.", new Object[]{identifier});
+                return true;
+            }
+
+            if (entity.getTimestamp() > alreadyListedEntity.getTimestamp()) {
+                logger.trace("Picked {} having newer timestamp {} than {}.",
+                        new Object[]{identifier, entity.getTimestamp(), alreadyListedEntity.getTimestamp()});
+                return true;
+            }
+
+            if (entity.getSize() != alreadyListedEntity.getSize()) {
+                logger.trace("Picked {} having different size {} than {}.",
+                        new Object[]{identifier, entity.getSize(), alreadyListedEntity.getSize()});
+                return true;
+            }
+
+            logger.trace("Skipped {}, not changed.", new Object[]{identifier, entity.getTimestamp(), minTimestampToList});
+            return false;
+        }).collect(Collectors.toList());
+
+        // Find old enough entries.
+        final List<String> oldEntityIds = alreadyListedEntities.entrySet().stream()
+                .filter(entry -> entry.getValue().getTimestamp() < minTimestampToList).map(Map.Entry::getKey)
+                .collect(Collectors.toList());
+
+        if (updatedEntities.isEmpty() && oldEntityIds.isEmpty()) {
+            logger.debug("None of updated or old entity was found. Yielding.");
+            context.yield();
+            return;
+        }
+
+        // Remove old entries.
+        oldEntityIds.forEach(oldEntityId -> alreadyListedEntities.remove(oldEntityId));
+
+        // Emit updated entities.
+        updatedEntities.forEach(updatedEntity -> {
```

Comment:
This again feels like a spot where imperative coding style is more natural than lambdas, given that it's a simple for-each with a significantly sized 'body'. The lambda here doesn't really buy anything but leads to much more cryptic stack traces and is more expensive to evaluate.

Reply:
Done.